src/lib/tuning/abstract_selector.hpp

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "tuning/tuning_choice.hpp"
+#include "tuning/tuning_operation.hpp"
+
+namespace opossum {
+
+/**
+ * An AbstractSelector transforms an unordered list of TuningChoices into
+ * a concrete sequence of TuningOperations.
+ *
+ * It considers a cost budget that the entire operation sequence as well as any
+ * continuous subsequence from the beginning must not exceed.
+ * The operation sequence is prioritized by expected performance impact on the system,
+ * so that the most beneficial operations come before less useful operations.
+ *
+ * The underlying problem relates to the knapsack problem, but goes beyond that
+ * in scope (e.g. choices invalidating other choices).
+ * For more information, see e.g. Wikipedia:
+ *   https://en.wikipedia.org/wiki/Knapsack_problem
+ */
+class AbstractSelector {
+ public:
+  virtual ~AbstractSelector() {}
+  /**
+   * Determine the index operation sequence as specified above based on the
+   * given list of choices and the budget value.
+   */
+  virtual std::vector<std::shared_ptr<TuningOperation>> select(
```

Comment:
* How about `pick_choices`?
 * Also, why is this ordered? You have already thrown out options that you don‘t like, so why is the order important?
 * Why do you modify the vector instead of creating a new one?

Reply:
(2) As the class comment states: The TuningOperations are prioritized by maximum desirability to ensure a greatest possible desirability gain even if the execution of TuningOperations should be interrupted by a timeout.
(1) Therefore `pick_choices` would not reflect the important prioritization aspect. What about `pick_and_prioritize_choices` or `pick_most_desirable_choices`. But then we are not sure if it is not better to trade a more concise name for the necessity to read the comments.
(3) We _do_ create a new vector of TuningOperation-shared_ptrs.

Reply:
CC @mrzzzrm

Reply:
I understood two and three. Personally, I still don't like `select` too much, but I could live with it.